features/nvstore/source/nvstore.h

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright (c) 2016 ARM Limited. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __NVSTORE_H
+#define __NVSTORE_H
+
+#include <stdint.h>
+#include <stdio.h>
+#include "nvstore_int_flash_wrapper.h"
+#include "nvstore_shared_lock.h"
+
+#ifdef NVSTORE_TESTING
+#undef NVSTORE_PROBE_ONLY
+#endif
+
+enum {
+    NVSTORE_SUCCESS                =  0,
+    NVSTORE_READ_ERROR             = -1,
+    NVSTORE_WRITE_ERROR            = -2,
+    NVSTORE_NOT_FOUND              = -3,
+    NVSTORE_DATA_CORRUPT           = -4,
+    NVSTORE_BAD_VALUE              = -5,
+    NVSTORE_BUFF_TOO_SMALL         = -6,
+    NVSTORE_FLASH_AREA_TOO_SMALL   = -7,
+    NVSTORE_OS_ERROR               = -8,
+    NVSTORE_BUFF_NOT_ALIGNED       = -9,
+    NVSTORE_ALREADY_EXISTS         = -10,
+};
+
+#ifndef NVSTORE_MAX_KEYS
+#define NVSTORE_MAX_KEYS 16
+#endif
+
+class NVStore {
+public:
+
+/**
+ * @brief As a singleton, return the single instance of the class.
+ *        Reason for this class being a singleton is the following:
+ *        - Ease the use for users of this class not having to coordinate instantiations.
+ *        - Lazy instantiation of internal data (which we can't achieve with simple static classes).
+ *
+ * @returns Singleton instance reference.
+ */
+    static NVStore& get_instance()
+    {
+        // Use this implementation of singleton (Meyer's) rather than the one that allocates
+        // the instance on the heap, as it ensures destruction at program end (preventing warnings
+        // from memory checking tools such as valgrind).
+        static NVStore instance;
```

Comment:
thread safe for the C++ standard we use at the moment?

Reply:
Yes - this implementation is thread safe. It's similar to the static implementation tested in TESTS/mbed_drivers/race_test.

Reply:
Meyer's singleton is thread safe in C++11 and later revisions of the language; not C++98/03.

Reply:
Indeed so. However, as far as I managed to understand, this specific feature (accessing static member for the first time by multiple threads) has been verified to work in our currently supported tool chains. I don't mind adding a lock there, but this would affect performance (each time get_instance is invoked).
@c1728p9, @geky - care to elaborate about this?

Reply:
Anyway, I can add a lock there the same way as SingletonPtr does.

Reply:
We have quite a bit of code relying on this being thread-safe, and I believe/hope it is with all our toolchains.

Technically nothing's thread-safe in C++98/03, because they don't have threads.

Reply:
We  provide guards for ARMCC and GCC, not certain about IAR but should (they mention guard calls and multithreaded lib support that we both enabled).

should be fine as it is

Reply:
@kjbracey-arm Out of curiosity I've made a quick check. It is safe with all the compilers we support.

Reply:
As @davidsaada mentioned, there's also a test specificaly to check this:
https://github.com/ARMmbed/mbed-os/blob/master/TESTS/mbed_drivers/race_test/main.cpp#L50

Reply:
So right now I'm keeping this as is.